src/Ast.ml

```diff
@@ -217,8 +217,67 @@ let longident_is_simple c x =
 let module_type_is_simple x =
   match x.pmty_desc with Pmty_signature _ -> false | _ -> true
 
-let module_expr_is_simple x =
-  match x.pmod_desc with Pmod_structure _ -> false | _ -> true
+let rec module_expr_is_simple x =
+  match x.pmod_desc with
+  | Pmod_structure _ -> false
+  | Pmod_constraint (e, t) ->
+      module_expr_is_simple e && module_type_is_simple t
+  | _ -> true
+
+let signature_item_is_simple x =
+  match x.psig_desc with
+  | Psig_value _ -> false
+  | Psig_type (_, l) -> List.length l <= 1
+  | Psig_typext _ -> true
+  | Psig_exception _ -> true
+  | Psig_module {pmd_type= m} -> module_type_is_simple m
+  | Psig_recmodule _ -> false
+  | Psig_modtype {pmtd_type= m} -> Option.for_all m ~f:module_type_is_simple
+  | Psig_open _ -> true
+  | Psig_include _ -> true
+  | Psig_class _ -> false
+  | Psig_class_type _ -> false
+  | Psig_attribute _ -> true
+  | Psig_extension _ -> false
+
+let value_binding_is_simple c x =
+  match x.pvb_expr.pexp_desc with
+  | Pexp_ident l -> longident_is_simple c l.txt
+  | Pexp_constant _ -> true
+  | Pexp_let _ -> false
+  | Pexp_function [_] -> true
+  | Pexp_function _ -> false
+  | Pexp_fun _ -> false
```

Comment:
If you are allowing `function` with one pattern, why not allowing functions with simple body here ? (eg. calling recursively on the body)